### PR TITLE
sv-compositor: witness type zipped with timestamp

### PR DESCRIPTION
**Pull request by jleightcap**
bundle a timestamp into the witness structure when parsed.

### CODE REVIEW

File: sv_circuit/src/export.rs [hunk 1]
```diff
@@ -28,7 +28,7 @@ pub fn private<F: Write>(writer: &mut F, witness: &Witness) -> Result<()> {
     writeln!(writer, "@type field 2;")?;
     writeln!(writer, "@begin")?;
 
-    for (i, step) in witness.iter().enumerate() {
+    for (step, i) in witness.iter() {
         writeln!(writer, "// step {}", i)?;
```
Comment (jleightcap) [inline sv_circuit/src/export.rs:32]: here the timestamp isn't serialized into the private input, just kept around as part of the witness structure.

File: sv_circuit/src/parse.rs [hunk 2]
```diff
@@ -1,28 +1,36 @@
...
+            ))),
         })
         .collect::<Vec<_>>()
         .try_into()
-        .map_err(|v| anyhow!("bad witness step {:?}", v))
+        .map_err(|_| SVCircuitError::WitnessError("FUCK".to_string()))
 }
 
-pub fn witness(f: BufReader<File>) -> Result<Witness> {
-    f.lines().flat_map(|l| l.map(witness_line)).collect()
```
Comment (jleightcap) [inline sv_circuit/src/parse.rs:22]: the use of `flat_map` here just removes errors from the collected vector, the correct way of bailing on errors is `collect`ing `Vec<Result<_>>` into `Result<Vec<_>` and using `?`.

**Comment by jleightcap** [comment]
@lisaoverall for serialization, my idea is to append the timestamp on the _end_ of the execution trace, so it is included in each step but but not called as input to the tiny86 `@function`. This is a hack, but requires the minimal amount of effort. Let me know what you think.

**Comment by lisaoverall** [comment]
@jleightcap Appending the timestamp makes sense to me. Instead of the current `WITNESS_LEN` bits per step, each step will be as long as the witness + the timestep. We will just need to take care with things like loop bounds and uses of `circuit.inputs.len()` when exporting IR0 `@function`s to make sure we're eliding the timestamp bits for the execution consistency proof.

**Comment by jleightcap** [comment]
> @jleightcap Appending the timestamp makes sense to me. Instead of the current `WITNESS_LEN` bits per step, each step will be as long as the witness + the timestep. We will just need to take care with things like loop bounds and uses of `circuit.inputs.len()` when exporting IR0 `@function`s to make sure we're eliding the timestamp bits for the execution consistency proof.

Great, we're on the same page here. I'll build on this PR when timestamp serialization is needed in checks: this PR is ready to merge, then.

File: sv_circuit/src/parse.rs [hunk 3]
```diff
@@ -18,7 +18,7 @@ fn witness_line(step: String) -> Result<WitnessStep, SVCircuitError> {
         })
         .collect::<Vec<_>>()
         .try_into()
-        .map_err(|_| SVCircuitError::WitnessError("FUCK".to_string()))
```
Comment (jleightcap) [inline sv_circuit/src/parse.rs]: not the most descriptive error 🙃

**Review by lisaoverall — APPROVED** [review]

File: sv_circuit/src/parse.rs (hunk 3 above)
Comment (lisaoverall) [reply sv_circuit/src/parse.rs]: It's okay, I'm used to doing a "sanitize for senior leader consumption" pass. I reverted to the previous message.

**Pull request merged**